gumble/jitterbuffer.go

```diff
@@ -0,0 +1,202 @@
+package gumble
+
+import (
+	"container/heap"
+	"encoding/binary"
+	_ "fmt"
+	_ "log"
+	"math"
+	_ "os"
+	"sync"
+	"time"
+)
+
+const (
+	// JitterStartDelay is the starting delay we will start our buffer with
+	jitterStartDelay = 200 * time.Millisecond
+)
+
+// jbAudioPacket holds pre-decoded audio samples
+type jbAudioPacket struct {
+	Sequence int64
+	Client   *Client
+	Sender   *User
+	Target   *VoiceTarget
+	Samples  int
+	Opus     []byte
+	Length   int
+
+	HasPosition bool
+	X, Y, Z     float32
+	IsLast      bool
+}
+
+type jitterBufferHeap []*jbAudioPacket
+
+func (h jitterBufferHeap) Len() int           { return len(h) }
+func (h jitterBufferHeap) Less(i, j int) bool { return h[i].Sequence < h[j].Sequence }
+func (h jitterBufferHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
+
+func (h *jitterBufferHeap) Push(x interface{}) {
+	*h = append(*h, x.(*jbAudioPacket))
+}
+
+func (h *jitterBufferHeap) Pop() interface{} {
+	old := *h
+	x := old[len(old)-1]
+	*h = old[:len(old)-1]
+	return x
+}
+
+// jitterBuffer struct holds all information to run the jitter buffer
+// seq is the current sequence number of the audio packets we have sent to be played
+type jitterBuffer struct {
+	seq           int64
+	jitter        time.Duration
+	heap          jitterBufferHeap
+	bufferSamples int64
+	running       bool
+	user          *User
+	target        *VoiceTarget
+	client        *Client
+	HeapLock      *sync.Mutex
+	RunningLock   *sync.Mutex
+}
+
+// Creates a new jitterBuffer
+func newJitterBuffer() *jitterBuffer {
+	jb := &jitterBuffer{
+		running:     false,
+		seq:         -1,
+		jitter:      jitterStartDelay,
+		HeapLock:    &sync.Mutex{},
+		RunningLock: &sync.Mutex{},
+	}
+	return jb
+}
+
+// AddPacket adds a packet to the jitter buffer
+func (j *jitterBuffer) AddPacket(ap *jbAudioPacket) error {
+	samples, err := j.user.decoder.SampleSize(ap.Opus)
+	if err != nil {
+		return err
+	}
+	ap.Samples = samples
+	j.bufferSamples += int64(ap.Samples)
+	j.HeapLock.Lock()
+	heap.Push(&j.heap, ap)
+	j.HeapLock.Unlock()
+	if !j.running {
+		j.RunningLock.Lock()
+		j.running = true
+		j.RunningLock.Unlock()
+		if j.seq == -1 || len(j.heap) == 1 { // set our sequence to first received audio packet's sequence if buffer is empty (-1)
+			//println("Heap empty, or -1, sequence is now at ", ap.Sequence)
+			j.seq = ap.Sequence
+		}
+		j.client = ap.Client
+		go j.process()
+	}
+	return nil
+}
+
+// TODO don't depend on is last
+func (j *jitterBuffer) process() {
+	time.Sleep(j.jitter)
+	var chans []chan *AudioPacket
+	j.client.volatileLock.Lock()
+	j.client.volatileWg.Wait()
+	for item := j.client.Config.AudioListeners.head; item != nil; item = item.next {
+		j.client.volatileLock.Unlock()
+		ch := make(chan *AudioPacket)
+		defer close(ch)
+		chans = append(chans, ch)
+		event := AudioStreamEvent{
+			Client: j.client,
+			User:   j.user,
+			C:      ch,
+		}
+		item.listener.OnAudioStream(&event)
```

Comment:
This bit is broken. Creating the `AudioPacket` channel and calling `OnAudioStream()` should only happen once. The client has a goroutine running listening on the channel, so it doesn't make sense to be closing it every time the heap size goes to zero (which happens when nobody is speaking).